### PR TITLE
[SPARK-25358][SQL] MutableProjection supports fallback to an interpreted mode

### PR DESCRIPTION
**Pull request by maropu**
## What changes were proposed in this pull request?
In SPARK-23711, `UnsafeProjection` supports fallback to an interpreted mode. Therefore, this pr fixed code to support the same fallback mode in `MutableProjection` based on `CodeGeneratorWithInterpretedFallback`.

## How was this patch tested?
Added tests in `CodeGeneratorWithInterpretedFallbackSuite`.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Projection.scala [hunk 1]
```diff
@@ -178,15 +236,7 @@ object UnsafeProjection
...
-    val unsafeExprs = toUnsafeExprs(toBoundExprs(exprs, inputSchema))
-    try {
-      GenerateUnsafeProjection.generate(unsafeExprs, subexpressionEliminationEnabled)
-    } catch {
-      case NonFatal(_) =>
-        // We should have already seen the error message in `CodeGenerator`
-        logWarning("Expr codegen error and falling back to interpreter mode")
-        InterpretedUnsafeProjection.createProjection(unsafeExprs)
-    }
+    createObject(toUnsafeExprs(toBoundExprs(exprs, inputSchema)), subexpressionEliminationEnabled)
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Projection.scala]: Removed the duplicated fallback logic here and reused the logic of `CodeGeneratorWithInterpretedFallback`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/InterpretedMutableProjection.scala [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+      i += 1
+    }
+    i = 0
+    while (i < exprArray.length) {
+      mutableRow(i) = buffer(i)
+      i += 1
+    }
+    mutableRow
+  }
+}
```
Comment (sadhen) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/InterpretedMutableProjection.scala]: This is my improved version of InterpretedMutableProject:

```
  override def apply(input: InternalRow): InternalRow = {
    var i = 0
    while (i < exprArray.length) {
      // Store the result into buffer first, to make the projection atomic (needed by aggregation)
      if (exprArray(i) != NoOp) {
        buffer(i) = exprArray(i).eval(input)
      }
      i += 1
    }
    i = 0
    while (i < exprArray.length) {
      if (exprArray(i) != NoOp) {
        mutableRow(i) = buffer(i)
      }
      i += 1
    }
    mutableRow
  }
```

The AggregationIterator uses NoOp. If you replace the codegen one with the interpreted one. You will encounter an exception from `NoOp.eval`.

Comment (sadhen) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/InterpretedMutableProjection.scala]: The improvement (written 3 months ago) is based on the generated Java code.

**Comment by kiszk** [comment]
retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/InterpretedMutableProjection.scala (hunk 2 above)
Comment (kiszk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/InterpretedMutableProjection.scala]: Looks good point. Do you know whether there is a test cause that causes the exception with the interpreted one?

Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/InterpretedMutableProjection.scala]: The change seems to look good to me though, I'ld like to address performance and code quality issues in follow-ups. cc: @gatorsmile

**Comment by maropu** [comment]
Can you check? @gatorsmile @rednaxelafx @cloud-fan @viirya

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/InterpretedMutableProjection.scala (hunk 2 above)
Comment (sadhen) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/InterpretedMutableProjection.scala]: Should be easy to create a test case.

Comment (rednaxelafx) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/InterpretedMutableProjection.scala]: +1 on the check for `NoOp`s.

**Review by rednaxelafx — COMMENTED** [review]
Looks good to me except for @sadhen's comment about checking `NoOp`. I'm good with this PR once that's addressed.

Thank you so much for handling the mutable projection case as well. That'd help with cases like very wide `SortAggregate` a lot.

**Comment by maropu** [comment]
Thanks for your review, kirs! I'll update in a day.

**Comment by sadhen** [comment]
@maropu there are tips to view the codegen:

> spark.sql("explain codegen select 1 + 1").show(false)

You may explain an aggregation, and the NoOp check is obvious.

**Comment by maropu** [comment]
@sadhen aha, I see. yea, we might need to address it in this pr.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/InterpretedMutableProjection.scala [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+/**
+ * A [[MutableProjection]] that is calculated by calling `eval` on each of the specified
+ * expressions.
+ *
+ * @param expressions a sequence of expressions that determine the value of each column of the
+ *                    output row.
+ */
+class InterpretedMutableProjection(expressions: Seq[Expression]) extends MutableProjection {
+  def this(expressions: Seq[Expression], inputSchema: Seq[Attribute]) =
+    this(expressions.map(toBoundExprs(_, inputSchema)))
```
Comment (rednaxelafx) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/InterpretedMutableProjection.scala]:
```
[error] /home/jenkins/workspace/SparkPullRequestBuilder@2/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/InterpretedMutableProjection.scala:32: type mismatch;
[error]  found   : org.apache.spark.sql.catalyst.expressions.Expression
[error]  required: Seq[org.apache.spark.sql.catalyst.expressions.Expression]
[error]     this(expressions.map(toBoundExprs(_, inputSchema)))
[error]                                       ^
```

It's probably `this(toBoundExprs(expressions, inputSchema))` right?

Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/InterpretedMutableProjection.scala]: yea, haha. I've already updated!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/CodeGeneratorWithInterpretedFallback.scala [hunk 4]
```diff
@@ -37,19 +37,22 @@ object CodegenObjectFactoryMode extends Enumeration {
  */
 abstract class CodeGeneratorWithInterpretedFallback[IN, OUT] extends Logging {
 
-  def createObject(in: IN): OUT = {
+  def createObject(in: IN): OUT =
+    createObject(in, subexpressionEliminationEnabled = false)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/CodeGeneratorWithInterpretedFallback.scala]: Can we eliminate it? I think we can use `SQLConf.get` directly when we need to access the conf. This should be done in a different PR though.

Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/CodeGeneratorWithInterpretedFallback.scala]: ah, it might be yes. I'll check in follow-up.

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/CodeGeneratorWithInterpretedFallback.scala]: can we do that first? I think that will be a small change and can also make this PR simpler and easier to review.

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/CodeGeneratorWithInterpretedFallback.scala]: I took some more look, seems it's not an actual config, but just a method parameter. Maybe we should remove this config in 3.0.

that said, let's keep your PR as it is.

Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/CodeGeneratorWithInterpretedFallback.scala]: ok, I'll do that first in another pr.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/InterpretedMutableProjection.scala [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+  private[this] var mutableRow: InternalRow = new GenericInternalRow(expressions.size)
+  def currentValue: InternalRow = mutableRow
+
+  override def target(row: InternalRow): MutableProjection = {
+    mutableRow = row
+    this
+  }
+
+  override def apply(input: InternalRow): InternalRow = {
+    validExprs.foreach { case (expr, i) =>
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/InterpretedMutableProjection.scala]: Can you please use the old code? That should be much more performant that this.

Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/InterpretedMutableProjection.scala]: oh, I forgot that we should do that in performance-sensitive places... Thanks! I'll update.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CodeGeneratorWithInterpretedFallbackSuite.scala [hunk 6]
```diff
@@ -69,11 +85,26 @@ class CodeGeneratorWithInterpretedFallbackSuite extends SparkFunSuite with PlanT
   test("codegen failures in the CODEGEN_ONLY mode") {
     val errMsg = intercept[ExecutionException] {
       val input = Seq(BoundReference(0, IntegerType, nullable = true))
-      val codegenOnly = CodegenObjectFactoryMode.CODEGEN_ONLY.toString
       withSQLConf(SQLConf.CODEGEN_FACTORY_MODE.key -> codegenOnly) {
         FailedCodegenProjection.createObject(input)
       }
+      val noCodegen = CodegenObjectFactoryMode.NO_CODEGEN.toString
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CodeGeneratorWithInterpretedFallbackSuite.scala]: unnecessary line?

Comment (maropu) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CodeGeneratorWithInterpretedFallbackSuite.scala]: oh.. yes. I'll remove.

**Comment by cloud-fan** [comment]
LGTM.

**Comment by cloud-fan** [comment]
So the next step is, create a `Predicate` factory with fallback?

**Comment by maropu** [comment]
yea, we should do. Before that, I was thinking I would work on fallback for 
`SafeProjection` (https://issues.apache.org/jira/browse/SPARK-25374).

**Comment by maropu** [comment]
cc: @gatorsmile @rednaxelafx Could you check again and merge?

**Comment by rednaxelafx** [comment]
LGTM as well.

**Comment by maropu** [comment]
ping

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by maropu** [comment]
Thanks for the merging!
